### ISSUE TITLE
Missing number in PDF hyperlinks to code listings

### ISSUE THREAD

**Issue by jfbu** [sphinx-doc/sphinx#2414]
Consider this `index.rst`:

``` rst
The first hyperlink :numref:`toolbox.code_sdo1` should read "Listing 1" and
this other hyperlink :numref:`toolbox.code_sdo2` should read "Listing 2" but
the 1 and 2 are missing. (besides the link jump to bottom not top of code
snippet, this has already been reported at `#2405`)

.. literalinclude:: sdo1.m
   :name: toolbox.code_sdo1
   :caption: Code snippet

.. literalinclude:: sdo2.m
   :name: toolbox.code_sdo2
   :caption: Other code

```

with files `sdo1.m` and `sdo2.m` of random contents.

```
foo
foo
foo
```

and 

```
bar
bar
bar
```

Then the pdf shows as:

![capture d ecran 2016-04-04 a 09 49 16](https://cloud.githubusercontent.com/assets/2589111/14241304/0e78a57e-fa4b-11e5-87e4-dac48e24bfb9.png)

The https://github.com/jfbu/sphinx/tree/jumptorightplace-onstable fixes this issue, but needs reviewing because I don't know enough functioning of Sphinx LaTeX writer to see if it can break something else.

Notice that #2405 is also fixed by https://github.com/jfbu/sphinx/tree/jumptorightplace-onstable.

**Comment by jfbu** [comment]
Fixed by PR #2415 which needs reviewing.

**Comment by tk0miya** [comment]
Fixed by #2415 :-)
